src/device/v4l2/stateless/request.rs

```diff
@@ -0,0 +1,206 @@
+use v4l2r::controls::SafeExtControl;
+use v4l2r::ioctl;
+
+use crate::device::v4l2::stateless::device::V4l2Device;
+use crate::device::v4l2::stateless::queue::V4l2CaptureBuffer;
+use crate::device::v4l2::stateless::queue::V4l2OutputBuffer;
+
+struct InitRequestHandle {
+    device: V4l2Device,
+    timestamp: u64,
+    handle: ioctl::Request,
+    buffer: V4l2OutputBuffer,
+}
+
+impl InitRequestHandle {
+    fn new(
+        device: V4l2Device,
+        timestamp: u64,
+        handle: ioctl::Request,
+        buffer: V4l2OutputBuffer,
+    ) -> Self {
+        Self {
+            device,
+            timestamp,
+            handle,
+            buffer,
+        }
+    }
+    fn ioctl<C, T>(&mut self, ctrl: C) -> &mut Self
+    where
+        C: Into<SafeExtControl<T>>,
+        T: ExtControlTrait,
+    {
+        let which = ioctl::CtrlWhich::Request(self.handle.as_raw_fd());
+        let mut ctrl: SafeExtControl<T> = ctrl.into();
+        ioctl::s_ext_ctrls(&self.device, which, &mut ctrl).expect("Failed to set output control");
+        self
+    }
+    fn write(&mut self, data: &[u8]) -> &mut Self {
+        self.buffer.write(data);
+        self
+    }
+    fn submit(self) -> PendingRequestHandle {
+        self.buffer.submit(self.timestamp, self.handle.as_raw_fd());
+        self.handle.queue().expect("Failed to queue request handle");
+        PendingRequestHandle {
+            device: self.device.clone(),
+            timestamp: self.timestamp,
+        }
+    }
+}
+
+struct PendingRequestHandle {
+    device: V4l2Device,
+    timestamp: u64,
+}
+
+impl PendingRequestHandle {
+    fn sync(self) -> DoneRequestHandle {
+        DoneRequestHandle {
+            buffer: Rc::new(RefCell::new(self.device.sync(self.timestamp))),
+        }
+    }
+}
+
+struct DoneRequestHandle {
+    buffer: Rc<RefCell<V4l2CaptureBuffer>>,
+}
+
+impl DoneRequestHandle {
+    fn result(&self) -> V4l2Result {
+        V4l2Result {
+            buffer: self.buffer.clone(),
+        }
+    }
+}
+
+#[derive(Default)]
+enum RequestHandle {
+    Init(InitRequestHandle),
+    Pending(PendingRequestHandle),
+    Done(DoneRequestHandle),
+    #[default]
+    Unknown,
+}
+
+impl RequestHandle {
+    fn new(
+        device: V4l2Device,
+        timestamp: u64,
+        handle: ioctl::Request,
+        buffer: V4l2OutputBuffer,
+    ) -> Self {
+        Self::Init(InitRequestHandle::new(device, timestamp, handle, buffer))
+    }
+    fn timestamp(&self) -> u64 {
+        match self {
+            Self::Init(handle) => handle.timestamp,
+            Self::Pending(handle) => handle.timestamp,
+            Self::Done(handle) => handle.buffer.borrow().timestamp(),
+            _ => panic!("ERROR"),
+        }
+    }
+    fn ioctl<C, T>(&mut self, ctrl: C) -> &mut Self
+    where
+        C: Into<SafeExtControl<T>>,
+        T: ExtControlTrait,
+    {
+        match self {
+            Self::Init(handle) => handle.ioctl(ctrl),
+            _ => panic!("ERROR"),
```

Comment:
In order to avoid these runtime panics, you can use something called the typestate pattern to ensure that all state transitions of `RequestHandle` are valid at compile-time: https://cliffle.com/blog/rust-typestate/

This also makes me think that maybe your `Request` type could be useful beyond the stateless decoder. Let's see how this turns out, but maybe we can move it into `v4l2r`  to make it useful for other use-cases as well.